satellite/satellitedb/orders.go

```diff
@@ -214,3 +218,178 @@ func (db *ordersDB) UnuseSerialNumber(ctx context.Context, serialNumber storj.Se
+		if err != nil {
+			return nil, err
+		}
+		_, err = tx.Exec(db.db.Rebind(insert), request.OrderLimit.StorageNodeId.Bytes(), request.OrderLimit.SerialNumber.Bytes())
+		if err != nil {
+			if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
+				reject(request.OrderLimit.SerialNumber)
+			} else {
+				return nil, Error.Wrap(err)
+			}
+		}
+	}
+
+	// call to get all the bucket IDs
+	query := db.buildGetBucketIdsQuery(len(requests))
+	statement := db.db.Rebind(query)
+
+	args := make([]interface{}, len(requests))
+	for i, request := range requests {
+		args[i] = request.OrderLimit.SerialNumber.Bytes()
+	}
+
+	rows, err := tx.Query(statement, args...)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	bucketMap := make(map[storj.SerialNumber][]byte)
+	for rows.Next() {
+		var serialNumber, bucketID []byte
+		err := rows.Scan(&serialNumber, &bucketID)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+		sn, err := storj.SerialNumberFromBytes(serialNumber)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+		bucketMap[sn] = bucketID
+	}
+
+	// build all the bandwidth updates into one sql statement
+	var updateRollupStatement string
+	for _, request := range requests {
+		_, rejected := rejectedRequests[request.OrderLimit.SerialNumber]
+		if rejected {
+			continue
+		}
+		bucketID, ok := bucketMap[request.OrderLimit.SerialNumber]
+		if !ok {
+			reject(request.OrderLimit.SerialNumber)
+			continue
+		}
+		projectID, bucketName, err := orders.SplitBucketID(bucketID)
+		if err != nil {
+			return nil, errs.Wrap(err)
+		}
+
+		stmt, err := db.buildUpdateBucketBandwidthRollupStatements(request.OrderLimit, request.Order, projectID[:], bucketName, intervalStart)
+		if err != nil {
+			return nil, err
+		}
+		updateRollupStatement += stmt
+
+		stmt, err = db.buildUpdateStorageNodeBandwidthRollupStatements(request.OrderLimit, request.Order, intervalStart)
+		if err != nil {
+			return nil, err
+		}
+		updateRollupStatement += stmt
+	}
+
+	_, err = tx.Exec(updateRollupStatement)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+	for _, request := range requests {
+		_, rejected := rejectedRequests[request.OrderLimit.SerialNumber]
+		if !rejected {
+			r := &pb.SettlementResponse{
+				SerialNumber: request.OrderLimit.SerialNumber,
+				Status:       pb.SettlementResponse_ACCEPTED,
+			}
+
+			responses = append(responses, r)
+		}
+	}
+	return responses, nil
+}
+
+func (db *ordersDB) buildGetBucketIdsQuery(argCount int) string {
+	args := make([]string, argCount)
+	for i := 0; i < argCount; i++ {
+		args[i] = "?"
+	}
+	return fmt.Sprintf("SELECT serial_number, bucket_id FROM serial_numbers WHERE serial_number IN (%s);\n", strings.Join(args, ","))
+}
+
+func (db *ordersDB) buildUpdateBucketBandwidthRollupStatements(orderLimit *pb.OrderLimit, order *pb.Order, projectID []byte, bucketName []byte, intervalStart time.Time) (string, error) {
+	hexName, err := db.toHex(bucketName)
```

Comment:
Why do they need to be converted to hex?

Reply:
I had to create raw SQL for these because the postgres driver doesn't like multiple parametrized statements in the same exec.